src/server/frontend_wayland/wl_pointer.cpp

```diff
@@ -263,33 +263,51 @@ void mf::WlPointer::axis(MirPointerEvent const* event)
     auto const v_scroll_stop = mir_pointer_event_axis_stop(event, mir_pointer_axis_vscroll);
     auto const h_scroll_discrete = mir_pointer_event_axis_value(event, mir_pointer_axis_hscroll_discrete);
     auto const v_scroll_discrete = mir_pointer_event_axis_value(event, mir_pointer_axis_vscroll_discrete);
+    auto const h_scroll_value120 = mir_pointer_event_axis_value(event, mir_pointer_axis_hscroll_value120);
+    auto const v_scroll_value120 = mir_pointer_event_axis_value(event, mir_pointer_axis_vscroll_value120);
     auto const axis_source = wayland_axis_source(mir_pointer_event_axis_source(event));
 
     // Don't send an axis source unless we have one and we're also sending some sort of axis event.
     if (axis_source &&
         version_supports_axis_source() &&
-        (h_scroll || v_scroll || h_scroll_stop || v_scroll_stop || h_scroll_discrete || v_scroll_discrete))
+        (h_scroll || v_scroll || h_scroll_stop || v_scroll_stop || h_scroll_value120 || v_scroll_value120))
     {
         send_axis_source_event(axis_source.value());
         needs_frame = true;
     }
 
-    if (h_scroll_discrete && version_supports_axis_discrete())
+    if (h_scroll_value120)
     {
-        send_axis_discrete_event(Axis::horizontal_scroll, h_scroll_discrete);
-        needs_frame = true;
+        if (version_supports_axis_value120())
+        {
+            send_axis_value120_event(Axis::horizontal_scroll, h_scroll_value120);
+            needs_frame=true;
+        }
+        else if (h_scroll_discrete && version_supports_axis_discrete())
+        {
+            send_axis_discrete_event(Axis::horizontal_scroll, h_scroll_discrete);
+            needs_frame = true;
+        }
     }
 
-    if (v_scroll_discrete && version_supports_axis_discrete())
+    if (v_scroll_value120)
     {
-        send_axis_discrete_event(Axis::vertical_scroll, v_scroll_discrete);
-        needs_frame = true;
+        if (version_supports_axis_value120())
+        {
+            send_axis_value120_event(Axis::vertical_scroll, v_scroll_value120);
+            needs_frame=true;
+        }
+        else if (v_scroll_discrete && version_supports_axis_discrete())
+        {
+            send_axis_discrete_event(Axis::vertical_scroll, v_scroll_discrete);
+            needs_frame = true;
+        }
     }
 
-    // If we sent discrete scroll events, we're required to also send corresponding non-discrete ones (even if their
-    // values are 0 for some reason)
+    // If we sent value120 scroll events, we're required to also send a non-value120 event on the same axis, evin if the
+    // value is 0.
```

Comment:
Also, is it the case that non-zero `discrete` implies non-zero `value120`?

Reply:
> While the axis_value120 event description states that this replaces the axis_discrete event, it does not say that an axis event is also sent

Since the discrete event requires an axis event, and the value120 replaces discrete with no mention either of requiring or not requiring it, I think it's at least ambiguous. Since there's no harm in sending it, might as well.

Reply:
The reason is likely because the .axis is what carries the timestamp.